File: 3. Longest Substring Without Repeating Characters/Main.java
import java.util.HashSet;
import java.util.Set;

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.lengthOfLongestSubstring("abcabcbb"));
    }
}

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLength = 0;
        
        char[] charArray = s.toCharArray();
        Set<Character> charSet = new HashSet<>();
        
        int i = 0;

        for (int j = 0; j < charArray.length; j++) {
            while (charSet.contains(charArray[j])) {
                charSet.remove(charArray[i]);
                i++;
            }
            charSet.add(charArray[j]);
            maxLength = Math.max(maxLength, j - i + 1);
        }

        return maxLength;
    }
}

File: 3. Longest Substring Without Repeating Characters/main.go
package main

import "fmt"

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
}

func lengthOfLongestSubstring(s string) int {
	var maxLength int

	charSet := map[byte]bool{}

	for i, j := 0, 0; j < len(s); j++ {
		for charSet[s[j]] {
			delete(charSet, s[i])
			i++
		}
		charSet[s[j]] = true
		maxLength = max(maxLength, j-i+1)
	}

	return maxLength
}

func max(a, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}
